src/util/result.h

```diff
@@ -31,54 +144,89 @@ struct Error {
+
+    template <typename OT, typename OF>
+    void MoveConstruct(Result<OT, OF>& other)
     {
-        assert(has_value());
-        return std::get<1>(m_variant);
+        *this << std::move(other);
+        if (other) this->MoveValue(other); else if (other.m_info) this->m_info->failure = std::move(other.m_info->failure);
     }
-    template <class U>
-    T value_or(U&& default_value) const&
+
+    template <typename FT, typename FF>
+    friend class Result;
+
+public:
+    //! Constructors, destructor, and assignment operator.
+    template <typename... Args>
+    Result(Args&&... args)
     {
-        return has_value() ? value() : std::forward<U>(default_value);
+        Construct([this](auto&&... x) { this->ConstructValue(std::forward<decltype(x)>(x)...); }, std::forward<Args>(args)...);
     }
-    template <class U>
-    T value_or(U&& default_value) &&
+    template <typename OT, typename OF>
+    Result(Result<OT, OF>&& other) { MoveConstruct(other); }
+    Result& operator=(Result&& other)
     {
-        return has_value() ? std::move(value()) : std::forward<U>(default_value);
+        if (*this) this->DestroyValue(); else this->m_info->failure.reset();
+        MoveConstruct(other);
+        return *this;
+    }
+    ~Result() { if (*this) this->DestroyValue(); }
+
+    //! Operator moving warning and error messages from one result to another.
+    //! Only moves message strings, does not change success or failure values of
+    //! either Result object.
+    template <typename OT, typename OF>
+    Result<OT, OF>&& operator<<(Result<OT, OF>&& other)
+    {
+        if (other.m_info && !this->m_info) {
+            this->m_info.reset(new detail::ErrorInfo<F>{.errors = std::move(other.m_info->errors),
+                                                        .warnings = std::move(other.m_info->warnings)});
+        } else if (other.m_info && this->m_info) {
+            detail::MoveElements(other.m_info->errors, this->m_info->errors);
+            detail::MoveElements(other.m_info->warnings, this->m_info->warnings);
+        }
+        return std::move(other);
     }
-    explicit operator bool() const noexcept { return has_value(); }
-    const T* operator->() const LIFETIMEBOUND { return &value(); }
-    const T& operator*() const LIFETIMEBOUND { return value(); }
-    T* operator->() LIFETIMEBOUND { return &value(); }
-    T& operator*() LIFETIMEBOUND { return value(); }
 };
 
-template <typename T>
-bilingual_str ErrorString(const Result<T>& result)
-{
-    return result ? bilingual_str{} : std::get<0>(result.m_variant);
-}
+//! Helper methods to format error strings.
+bilingual_str ErrorString(const std::vector<bilingual_str>& errors);
+bilingual_str ErrorString(const std::vector<bilingual_str>& errors, const std::vector<bilingual_str>& warnings);
+template <typename T, typename F>
+bilingual_str ErrorString(const Result<T, F>& result) { return ErrorString(result.GetErrors(), result.GetWarnings()); }
```

Comment:
I didn't think about the footgun, that's a good point why we probably indeed won't want to add the other fn later. I still think it's a bit awkward that the function name doesn't entirely capture what the function is doing, but I think it's within reason and may be worth the trade-off for brevity (vs `ErrorWarningString()`. Not sure what I prefer, so feel free to ignore.

The new docstring is great, thanks!

Reply:
re: https://github.com/bitcoin/bitcoin/pull/25665#discussion_r954211552

> I still think it's a bit awkward that the function name doesn't entirely capture what the function is doing, but I think it's within reason and may be worth the trade-off for brevity (vs `ErrorWarningString()`. Not sure what I prefer, so feel free to ignore.

Could call it `util::MessageString(result)` instead of `util::ErrorString(result)`. I think the longer name would be ok too. I was pushing back more against changing the functionality than changing the name.

I do think any renaming should happen in a separate PR, before or after this one. The function already exists and is called in current code. This PR is backwards compatible just extends the Result API without requiring changes to existing code.

Reply:
> The function already exists and is called in current code.

Right, I did not consider that. Don't think it's worth the extra PR without more demand for it so I'm happy to just leave it at `ErrorString` until then.